File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public static int maxProduct(int[] arr) {
        int n = arr.length; // Size of the array

        int pre = 1, suff = 1;
        int ans = Integer.MIN_VALUE;

        for (int i = 0; i < n; i++) {
            // If pre or suff becomes zero, reset them to 1
            if (pre == 0) pre = 1;
            if (suff == 0) suff = 1;

            // Calculate prefix product
            pre *= arr[i];

            // Calculate suffix product
            suff *= arr[n - i - 1];

            // Update the answer with the maximum of the prefix or suffix product
            ans = Math.max(ans, Math.max(pre, suff));
        }

        return ans;
    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
int maxProduct(vector<int>& a) {
        double prefix = 1, suffix = 1, ans = INT_MIN;
        int n = a.size();
        for (int i = 0; i < n; i++) {
            prefix = (!prefix ? 1 : prefix);
            suffix = (!suffix ? 1 : suffix);
            prefix *= a[i];
            suffix *= a[n - i - 1];
            ans = max(ans, max(prefix, suffix));
        }
        return (int)ans;
    }
};